main/src/library/List.flix

```diff
@@ -373,42 +373,32 @@ namespace List {
     /// That is, the result is of the form: `f(x1, 0) :: f(x2, 1) :: ...`.
     ///
     @Time(time(f) * length(l)) @Space(space(f) * length(l))
-    pub def mapWithIndex(f: (a, Int32) -> b & ef, l: List[a]): List[b] & ef = mapWithIndexHelper(f, l, 0, ks -> ks)
-
-    ///
-    /// Helper function for `mapWithIndex`.
-    ///
-    def mapWithIndexHelper(f: (a, Int32) -> b & ef, l: List[a], i: Int32, k: List[b] -> List[b]): List[b] & ef = match l {
-        case Nil     => k(Nil)
-        case x :: xs => {
-            let y = f(x, i);
-            mapWithIndexHelper(f, xs, i+1, ys -> k(y :: ys))
-        }
-    }
+    pub def mapWithIndex(f: (a, Int32) -> b & ef, l: List[a]): List[b] & ef =
+        def loop(ll, i, k) = match ll {
+            case Nil     => k(Nil)
+            case x :: xs =>
+                let y = f(x, i);
+                loop(xs, i + 1, ys -> k(y :: ys))
+        };
+        loop(l, 0, identity)
 
     ///
     /// Returns the result of applying `f` to every element in `l` and concatenating the results.
     ///
     @Time(time(f) * length(l)) @Space(time(f) * length(l))
-    pub def flatMap(f: a -> List[b] & ef, l: List[a]): List[b] & ef = flatMapHelper(f, l, ks -> ks)
-
-    ///
-    /// Helper function for `flatMap`.
-    ///
-    def flatMapHelper(f: a -> List[b] & ef, l: List[a], k: List[b] -> List[b]): List[b] & ef = match l {
-        case Nil     => k(Nil)
-        case x :: xs => {
-            let y = f(x);
-            flatMapHelper(f, xs, ys -> k(y ::: ys))
-        }
-    }
+    pub def flatMap(f: a -> List[b] & ef, l: List[a]): List[b] & ef =
+        def loop(ll, acc) = match ll {
+            case Nil     => acc
+            case x :: xs => loop(xs, acc ::: f(x))
```

Comment:
Does this produce the elements in the right order? Maybe it does, just hard to see :)

Reply:
Appends the list produced by f(x) to the previous lists so it preserves the order (instead of prepending lists and then reversing). Also passes the tests :)